src/jdk.jlink/share/classes/jdk/tools/jlink/internal/ImageResourcesTree.java

```diff
@@ -97,215 +108,207 @@ private static String buildPath(Node item) {
         }
+        // Exception only used for program flow, not debugging.
+        @Override
+        public Throwable fillInStackTrace() {return this;}
     }
 
     /**
      * Tree of nodes.
      */
-    private static final class Tree {
+    // Visible for testing only.
+    static final class Tree {
+        private static final String PREVIEW_PREFIX = "META-INF/preview/";
 
         private final Map<String, Node> directAccess = new HashMap<>();
         private final List<String> paths;
         private final Node root;
-        private Node modules;
-        private Node packages;
+        private Node packagesRoot;
 
-        private Tree(List<String> paths) {
-            this.paths = paths;
+        // Visible for testing only.
+        Tree(List<String> paths) {
+            this.paths = paths.stream().sorted(Comparator.reverseOrder()).toList();
+            // Root node is not added to the directAccess map.
             root = new Node("", null);
             buildTree();
         }
 
         private void buildTree() {
-            modules = new Node("modules", root);
-            directAccess.put(modules.getPath(), modules);
-
-            Map<String, Set<String>> moduleToPackage = new TreeMap<>();
-            Map<String, Set<String>> packageToModule = new TreeMap<>();
-
-            for (String p : paths) {
-                if (!p.startsWith("/")) {
-                    continue;
-                }
-                String[] split = p.split("/");
-                // minimum length is 3 items: /<mod>/<pkg>
-                if (split.length < 3) {
-                    System.err.println("Resources tree, invalid data structure, "
-                            + "skipping " + p);
-                    continue;
-                }
-                Node current = modules;
-                String module = null;
-                for (int i = 0; i < split.length; i++) {
-                    // When a non terminal node is marked as being a resource, something is wrong.
+            Node modulesRoot = new Node("modules", root);
+            directAccess.put(modulesRoot.getPath(), modulesRoot);
+            packagesRoot = new Node("packages", root);
+            directAccess.put(packagesRoot.getPath(), packagesRoot);
+
+            // Map of dot-separated package names to module references (those
+            // in which the package appear). References are merged after to
+            // ensure each module name appears only once, but temporarily a
+            // module may have several entries per package (e.g. with-content,
+            // without-content, normal, preview-only etc..).
+            Map<String, Set<ModuleReference>> packageToModules = new TreeMap<>();
+            for (String fullPath : paths) {
+                try {
+                    processPath(fullPath, modulesRoot, packageToModules);
+                } catch (InvalidTreeException err) {
```

Comment:
Exception varialble "err" is ambiguous with System.err used below.
"ex" is more conventional for exceptions.